### PR TITLE
Remove top level `with lib;` and `with import ...` in `pkgs/top-level/`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is #208242. This is a pure refactor: there is, or ought to be, no functional change contained in this PR, even though https://github.com/NixOS/nixpkgs/pull/293835/commits/83339e48d0fdabc4590507329600195f01e4dcb1 drops a few internal names from `release-lib.nix`.

This PR changes the files that used top-level `with lib;` and `with import ...;` in the `pkgs/top-level/` directory.

Here's what remains:

```
$ rg --sort=path '^with' pkgs/top-level/
pkgs/top-level/aliases.nix
20:with self;

pkgs/top-level/all-packages.nix
11:with pkgs;

pkgs/top-level/emscripten-packages.nix
3:with pkgs;

pkgs/top-level/java-packages.nix
3:with pkgs;

pkgs/top-level/linux-kernels.nix
22:with linuxKernel;

pkgs/top-level/metrics.nix
3:with pkgs;

pkgs/top-level/perl-packages.nix
23:with self; {

pkgs/top-level/python-aliases.nix
3:with self;

pkgs/top-level/python2-packages.nix
7:with self; with super; {
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drupol** [comment]
How am I supposed all these changes?

**Comment by philiptaron** [comment]
@drupol, good question:
 
> How am I supposed [to review] all these changes?

Yeah, it's a little difficult through the web UI of GitHub. 

Here's how I concocted them. Maybe that'll help with review.

1. Whenever there is a `with <expr>` statement, whatever is in `<expr>` must be an attrset, and whatever values that attrset has are exposed to the scope that the `with` covers. If the with is at the top level, that's going to be the rest of the file, usually. 
2. To remove the `with`, two steps are required:
    1. Discover all the values from the `<expr>` attrset
    2. Find which of those are being used in the file.
3. In the case of `with lib`, we can find those values with [`nix eval --json --apply builtins.attrNames nixpkgs#lib`](https://gist.github.com/philiptaron/04326000d5ff20cb72c2805efd09701c).
4. I ran `rg -wof /tmp/lib.nix.names` on (say) [pkgs/top-level/kodi-packages.nix](https://github.com/NixOS/nixpkgs/pull/293835/commits/fec9460e07d797a6d77e6aaf89d5ee4bc8545299)
5. I took the names and I put them in a let block at the same level as the `with` used to be.
6. If there were any uses of `lib.` that corresponded with one of those names, I removed the `lib.`. This is not strictly necessary, but it's both nice -- there's one way of referring to the same symbol, rather than two ways -- and it's slightly more performant, [as I understand it](https://github.com/NixOS/nix/issues/9592).

For the users of `release-lib.nix`, I did similarly.

1. I generated the list of attributes coming from `release-lib.nix`.
    * `$ nix eval --impure --json --apply builtins.attrNames --expr 'import ./pkgs/top-level/release-lib.nix { supportedSystems = null; }' | jq -r '.[]'` yields [these names.](https://gist.github.com/philiptaron/f003e177099e864e4db616045deabbb5)
2. For each of the names, I checked whether there was any mention outside of `release-lib.nix` whatsoever. These names failed that test:
   1. _mapTestOnHelper
   2. crossMaintainers
   3. mkPkgsFor
3. I used the "explicit inherit" technique to just have exactly the set of items that ought to be exported be exported.
4. In the users, I created a variable for the `release-lib.nix` import, then did the same `rg -wof /tmp/release-lib.nix.names` on the file to see what names each file used.

For the top-level files, I did this:

```
nix-instantiate  --eval pkgs/top-level/release.nix --arg supportedSystems '[ "x86-64-linux" ]'
```

Finally, wherever there was a chance to run the file directly -- most of the release*.nix files say how to run them -- I did so, Ctrl-C'ing the process once I was sure it had gotten through evaluating the top-level file.

**Comment by AndersonTorres** [comment]
> How am I supposed all these changes?

The best thing I have done in this UI is to activate side-by-side view instead of the typical stacked diffs.
It helps a lot comparing deletions and substitutions!

File: pkgs/top-level/kodi-packages.nix [hunk 1]
```diff
@@ -34,170 +23,186 @@ let self = rec {
     in
       unique (modules ++ concatLists (catAttrs "requiredKodiAddons" modules));
 
-  # package update scripts
+  self = {
```
Comment (fricklerhandwerk) [inline pkgs/top-level/kodi-packages.nix:26]: The original self was a `rec`, and `kodi` is used in `kodiAddonFor`. Pretty sure this is a semantic change.

```suggestion
  self = rec {
```

**Review by fricklerhandwerk — APPROVED** [review]
Looks good except for that one thing.

Although I'd like to add for consideration, that locally qualified calls to `lib` have a crucial advantage over `let ... inherit`: they scale better because they don't require updates in two locations. If you remove the last instance of a name or add a new one, you have to go up and update the `let ... inherit` block. Unqualified calls make the code more concise, but brevity doesn't always improve readability. For instance, I'd strongly advise against unconditional eta-reduction because in my experience that adds cognitive overhead when reading code.

Removing top-of-file `with` is definitely an improvement to readability, regardless what it's replaced by, but we should at least ask around before doing wholesale changes. In particular I'd like to have @infinisil's and @roberth's opinion on this.

File: pkgs/top-level/kodi-packages.nix (hunk 1 above)
Comment (philiptaron) [reply pkgs/top-level/kodi-packages.nix:26]: I don't believe it is...

1. `kodi` comes from line 1 in both diffs.
2. It's bound into the attrset at line 16 left, line 31 right.
3. The same `kodi` is still bound into `hasKodiAddon` and `toKodiAddon`.

Unless there's some subtlety about recursive attrsets that I'm failing to understand, I _think_ it all checks out. Maybe something about the `callPackage` definition and splicing?

I can definitely put the `rec` back: I'm more interested in learning what's missing in my mental model than I am about this particular change being one way or another.

**Comment by philiptaron** [comment]
> I'd like to add for consideration, that locally qualified calls to `lib` have a crucial advantage over `let ... inherit`: they scale better because they don't require updates in two locations. If you remove the last instance of a name or add a new one, you have to go up and update the `let ... inherit` block. Unqualified calls make the code more concise, but brevity doesn't always improve readability. For instance, I'd strongly advise against unconditional eta-reduction because in my experience that adds cognitive overhead when reading code.
> 
> Removing top-of-file `with` is definitely an improvement to readability, regardless what it's replaced by, but we should at least ask around before doing wholesale changes. In particular I'd like to have @infinisil's and @roberth's opinion on this.

One of the reasons I'm doing PR-by-PR changes instead of [a tree-wide PR](https://github.com/NixOS/nixpkgs/pull/209325) is precisely to have these conversations and get rough agreement, so I'm delighted to have the conversation, especially at the beginning of the effort.

On the topic of `lib.thing` vs. `thing`.

The number one thing I find clarifying is that there's one way to access the same symbol. For some reason, this cuts through the thicket of aliasing for me in a way little else does. I also prefer a bunch of use statements at the top of a file in Rust to `crate::symbol1`, `crate::symbol2` for the same reason. As I read the file, usually from top to bottom, it's a statement that "in this house we use the following things".

In that way, it's similar to the argument list in a `callPackage`-style package: it's the necessary prelude to what comes next.

I'll note that the argument that "you have to change it in two places, and it's hard to remember to do so" applies equally to `callPackage`-style packages, where the dependencies are first mentioned at the top of the file, and are then used in the body, sometimes at a large remove. 

As in that case, the impact should the user fail to remove the extra `inherit`ed symbol is low: it's dead code, but not much of it, and not a real tripping hazard.

But I'm quite open to other points of view. In particular, `lib.thing` for very short names (looking at you, `lib.or`) might be more readable.

You did say one thing that I didn't picture a code example for:

> I'd strongly advise against unconditional eta-reduction

Could you say more about what that looks like?

**Comment by fricklerhandwerk** [comment]
> As I read the file, usually from top to bottom, it's a statement that "in this house we use the following things".

This is a valid argument for slurping code, but fails for reviewing code, where you usually get tiny slices with little context. Such code style decisions have to make a trade-off between use cases. And we'd have to figure out how first-time readers interact with the code. If you're fixing a bug, chances are you'll grep yourself into the middle of a large file rather than reading a library top to bottom.

Again, maybe I'm overthinking it, because simply stating the origin of symbols is already a huge improvement, and the rest may not matter by far as much.



> > I'd strongly advise against unconditional eta-reduction
> 
> Could you say more about what that looks like?

Usually you'd eta-reduce something like this:

```nix
{ g, h }:
let
  f = name: value: g (h name) value;
in
attrs: builtins.mapAttrs f attrs
```

to something like this:

```nix
{ g, h }:
let
  f = name: g (h name);
in
builtins.mapAttrs f
```

But especially when attribute names and values have domain-specific meaning and you'd give them relevant symbol names, the second version makes that additional information disappear, which would otherwise have helped understanding both the purpose and the structure of the code. But that was an irrelevant aside for the `with` and `let ... inherit` problem.

File: pkgs/top-level/kodi-packages.nix (hunk 1 above)
Comment (fricklerhandwerk) [reply pkgs/top-level/kodi-packages.nix:26]: > 1. `kodi` comes from line 1 in both diffs.

Ah, I missed that. Seems alright then.

**Review by fricklerhandwerk — APPROVED** [review]

**Review by AndersonTorres — APPROVED** [review]

**Review by infinisil — APPROVED** [review]
The one case where this doesn't work is when you have the same symbol in multiple sub-libraries, such as `lib.{path,lists,strings}.hasPrefix` or `lib.{lists,fileset}.toList`. But in such a case we can either be explicit by using the fully-qualified attribute, or if performance is very important one can do

```nix
let
  pathHasPrefix = lib.path.hasPrefix;
in
```

In general I would also like `lib` to move towards more qualification instead of having (almost) everything exposed under `lib.*`. So e.g. instead of `lib.attrNames`, use `lib.attrs.names`. This will remove a lot of potential conflicts in the `lib.*` namespace, but it will make the above issue a bit worse.

It would be great if we didn't have to worry about the performance of this, because then there wouldn't be a problem with having more qualified names.

In any case, for now this seems fine to me, though it would be great to also hear from @roberth, who introduced the `inherit` from lib pattern :)

**Comment by infinisil** [comment]
Actually let's just merge, this should be fine :rocket:

**Comment by roberth** [comment]
> @roberth, who introduced the `inherit` from lib pattern :)

I don't think I invented it; I just applied it in `lib` itself.

I feel ambivalent about more "qualified names", because we don't actually have a language module system such that "qualified names" are a thing and exactly equivalent to importing the thing directly.
Also the names of sublibraries can interfere:
```nix
withMetadata = fileset: fileset.unions [ ./package.yaml ./LICENSE fileset ];
```

The added structure can be nice, but also arbitrary. Where does `mapAttrsToList` go? If it's `attrsets` or `lists`, surely `listToAttrs` can't be in the same sublibrary, etc.

**Pull request merged**